giskard/scanner/performance/performance_bias_detector.py

```diff
@@ -303,36 +303,88 @@ def _is_unbalanced_target(classes: pd.Series):
+
+    ctable = [[slice_x_cnt, slice_y_cnt], [comp_x_cnt, comp_y_cnt]]
+
+    # if the slice size is too small, use Fisher's exact test, otherwise use a chi-square test
+    if min(min(row) for row in ctable) <= max_size_fisher:
+        logger.info("PerformanceBiasDetector: Fisher's exact test")
+        return scipy.stats.fisher_exact(ctable, alternative="two-sided")[1]
+    logger.info("PerformanceBiasDetector: chi-square test")
+    return scipy.stats.chi2_contingency(ctable, correction=False, lambda_="log-likelihood")[1]
+
+
+def _calculate_pvalue_from_permutation_test(
+    slice_dataset, comp_dataset, dataset, model, metric, perm_test_resamples=1000
+):
+    logger.info("PerformanceBiasDetector: permutation test")
+
+    def statistic(slice_ids, comp_ids):
+        perm_slice_dataset = Dataset(
+            dataset.df.loc[slice_ids],
+            target=dataset.target,
+            column_types=dataset.column_types.copy(),
+            validation=False,
+        )
+        perm_comp_dataset = Dataset(
+            dataset.df.loc[comp_ids],
+            target=dataset.target,
+            column_types=dataset.column_types.copy(),
+            validation=False,
+        )
+        return metric(model, perm_slice_dataset).value - metric(model, perm_comp_dataset).value
+
+    slice_ids = slice_dataset.df.index.values
+    comp_ids = comp_dataset.df.index.values
+    perm_test_result = scipy.stats.permutation_test(
+        (slice_ids, comp_ids),
+        statistic=statistic,
+        permutation_type="independent",
+        n_resamples=perm_test_resamples,
+        alternative="two-sided",
+    )
+    return perm_test_result.pvalue
+
+
+def _calculate_slice_metrics(
+    model, dataset, metric, slice_fn, with_pvalue=False, max_size_fisher=30, perm_test_resamples=1000
+):
+    slice_dataset = dataset.slice(slice_fn)
+    slice_metric = metric(model, slice_dataset)
 
     if not with_pvalue:
-        return sliced_dataset, slice_metric, None
+        return slice_dataset, slice_metric, None
 
     # Perform statistical tests
-    complementary_dataset = dataset.slice(lambda df: df[~df.index.isin(sliced_dataset.df.index)], row_level=False)
-    comp_metric = metric(model, complementary_dataset)
+    comp_dataset = dataset.slice(lambda df: df[~df.index.isin(slice_dataset.df.index)], row_level=False)
+    comp_metric = metric(model, comp_dataset)
 
     try:
         # If we have raw values for the metric, we perform a standard t-test
+        logger.info(f"PerformanceBiasDetector: metric name = {slice_metric.name}")
         if slice_metric.raw_values is not None:
+            logger.info("PerformanceBiasDetector: t-test")
             alternative = "less" if metric.greater_is_better else "greater"
             _, pvalue = scipy.stats.ttest_ind(
                 slice_metric.raw_values, comp_metric.raw_values, equal_var=False, alternative=alternative
             )
+        elif metric.name.lower() in ["accuracy", "precision", "recall"]:
```

Comment:
We should do this check differently. Let’s use some attribute on the metric object, like is_binary_metric or similar.

Reply:
maybe it's even better to let the Metric class the responsibility of calculating the contingency table entries